File: app/lib/shared/models/achievement_model.dart
import 'dart:ui';

class AchievementModel {
  final String id;
  final String key;
  final String name;
  final String description;
  final String? verseReference;
  final String? verseText;
  final String iconEmoji;
  final String rarity;
  final int pfReward;
  final int manaReward;
  final bool isActive;

  const AchievementModel({
    required this.id, required this.key, required this.name,
    required this.description, this.verseReference, this.verseText,
    required this.iconEmoji, required this.rarity,
    required this.pfReward, required this.manaReward, required this.isActive,
  });

  factory AchievementModel.fromJson(Map<String, dynamic> json) => AchievementModel(
    id: (json['_id'] ?? json['id'] ?? '') as String,
    key: (json['key'] ?? '') as String,
    name: (json['name'] ?? '') as String,
    description: (json['description'] ?? '') as String,
    verseReference: json['verse_reference'] as String?,
    verseText: json['verse_text'] as String?,
    iconEmoji: (json['icon_emoji'] ?? '🏆') as String,
    rarity: (json['rarity'] ?? 'common') as String,
    pfReward: ((json['pf_reward'] ?? json['xp_reward'] ?? 0) as num).toInt(),
    manaReward: ((json['mana_reward'] ?? 0) as num).toInt(),
    isActive: (json['is_active'] ?? true) as bool,
  );

  Color get rarityColor {
    switch (rarity) {
      case 'rare': return const Color(0xFF1D4ED8);
      case 'epic': return const Color(0xFF7C3AED);
      default: return const Color(0xFF6B7280);
    }
  }
}

class UserAchievementModel {
  final String id;
  final AchievementModel achievement;
  final DateTime unlockedAt;

  const UserAchievementModel({
    required this.id, required this.achievement, required this.unlockedAt,
  });

  factory UserAchievementModel.fromJson(Map<String, dynamic> json) =>
    UserAchievementModel(
      id: (json['_id'] ?? json['id'] ?? '') as String,
      achievement: AchievementModel.fromJson(json['achievement_id'] as Map<String, dynamic>),
      unlockedAt: DateTime.parse(json['unlocked_at'] as String),
    );
}

File: api/src/modules/gamification/achievement.model.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IAchievement extends Document {
  key: string;
  name: string;
  description: string;
  verse_reference?: string;
  verse_text?: string;
  icon_emoji: string;
  icon_url: string;

  trigger: {
    type: 'lesson_count' | 'streak_days' | 'trail_complete' | 'league_rank' | 'invite_count' | 'pf_total' | 'level' | 'perfect_lesson' | 'pf_earn' | 'trail_progress' | 'study_complete' | 'league_xp' | 'review_lesson' | 'streak_maintain';
    value: number;
  };

  rarity: 'common' | 'rare' | 'epic';
  pf_reward: number;
  mana_reward: number;

  // Fields merged from missions
  cycle?: 'one_time' | 'daily' | 'weekly';
  difficulty?: 'easy' | 'medium' | 'hard';
  is_premium: boolean;

  is_active: boolean;
  sort_order: number;

  createdAt: Date;
  updatedAt: Date;
}

const AchievementSchema = new Schema<IAchievement>(
  {
    key: { type: String, required: true, unique: true },
    name: { type: String, required: true },
    description: { type: String, required: true },
    verse_reference: { type: String },
    verse_text: { type: String },
    icon_emoji: { type: String, default: '🏆' },
    icon_url: { type: String },

    trigger: {
      type: { type: String, enum: ['lesson_count', 'streak_days', 'trail_complete', 'league_rank', 'invite_count', 'pf_total', 'level', 'perfect_lesson', 'pf_earn', 'trail_progress', 'study_complete', 'league_xp', 'review_lesson', 'streak_maintain'], required: true },
      value: { type: Number, required: true },
    },

    rarity: { type: String, enum: ['common', 'rare', 'epic'], default: 'common' },
    pf_reward: { type: Number, default: 0 },
    mana_reward: { type: Number, default: 0 },

    cycle: { type: String, enum: ['one_time', 'daily', 'weekly'], default: 'one_time' },
    difficulty: { type: String, enum: ['easy', 'medium', 'hard'] },
    is_premium: { type: Boolean, default: false },

    is_active: { type: Boolean, default: true },
    sort_order: { type: Number, default: 0 },
  },
  { timestamps: true }
);

AchievementSchema.index({ is_active: 1, sort_order: 1 });
AchievementSchema.index({ key: 1 }, { unique: true });
AchievementSchema.index({ 'trigger.type': 1 });

export const AchievementModel = mongoose.model<IAchievement>('Achievement', AchievementSchema);
